app/src/main/java/de/tum/in/tumcampusapp/utils/Utils.kt

```diff
@@ -0,0 +1,486 @@
+package de.tum.`in`.tumcampusapp.utils
+
+import android.annotation.TargetApi
+import android.app.Activity
+import android.content.Context
+import android.content.Intent
+import android.content.IntentFilter
+import android.content.pm.PackageManager
+import android.graphics.Bitmap
+import android.graphics.Canvas
+import android.os.BatteryManager
+import android.os.Build
+import android.preference.PreferenceManager
+import android.support.v4.content.ContextCompat
+import android.support.v4.content.pm.PackageInfoCompat
+import android.text.Html
+import android.text.Spanned
+import android.util.Log
+import android.widget.Toast
+import com.google.common.base.Charsets
+import com.google.gson.Gson
+import de.tum.`in`.tumcampusapp.BuildConfig
+import java.io.IOException
+import java.io.InputStream
+import java.io.PrintWriter
+import java.io.StringWriter
+import java.util.regex.Pattern
+
+/**
+ * Class for common helper functions used by a lot of classes.
+ */
+object Utils {
+    private const val LOGGING_REGEX = "[a-zA-Z0-9.]+\\."
+
+    @JvmStatic
+    fun migrateSharedPreferences(context: Context) {
+        val prefsLegacy = context.getSharedPreferences("internal_prefs", Context.MODE_PRIVATE)
+        val entries = prefsLegacy.all
+
+        if (entries.isEmpty()) {
+            return
+        }
+
+        val prefsEditor = PreferenceManager.getDefaultSharedPreferences(context).edit()
+
+        for ((key, value) in entries) {
+            when (value) {
+                is Boolean -> prefsEditor.putBoolean(key, value)
+                is String -> prefsEditor.putString(key, value)
+                is Int -> prefsEditor.putInt(key, value)
+                is Float -> prefsEditor.putFloat(key, value)
+                is Long -> prefsEditor.putLong(key, value)
+            }
+        }
+        prefsEditor.apply()
+
+        //Delete any old settingsPrefix
+        prefsLegacy.edit()
+                .clear()
+                .apply()
+    }
+
+    /**
+     * Get a value from the default shared preferences
+     *
+     * @param c          Context
+     * @param key        setting name
+     * @param defaultVal default value
+     * @return setting value, defaultVal if undefined
+     */
+    @JvmStatic
+    fun getSetting(c: Context, key: String, defaultVal: String): String {
+        val sp = PreferenceManager.getDefaultSharedPreferences(c)
+        return sp.getString(key, defaultVal)!!
```

Comment:
`SharedPreferences.getString()` is annotated as `@Nullable`, for the case there is a null default value, therefore we need to force unwrap.
Since we have a guaranteed non-null default value, the force unwrap is safe.

Reply:
so you suggest a `?: return defaultVal`?

Reply:
You can simply omit it.

Reply:
No, then I'm getting a big warning, that I return a nullable value from a non-nullable function. Definitely not doing that

Reply:
Odd, it is not giving me a warning and builds without any problems. Well, then I guess `!!` are fine